File: examples/java/src/main/java/com/example/locoaware/Repositories.java
package com.example.locoaware;

import com.fasterxml.jackson.databind.JsonNode;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * In-memory stand-in for a document store. The shape mirrors what you'd push
 * into MongoDB / DocumentDB / CosmosDB / Firestore / Couchbase. The point of
 * the examples is to show the SHAPE of the documents — the actual driver is
 * one swap away.
 *
 * <p>Production replacements (typical Spring Boot wiring):
 * <ul>
 *   <li>MongoDB     {@code spring-boot-starter-data-mongodb} — {@code MongoTemplate.upsert(query, update, "devices")}</li>
 *   <li>CosmosDB    {@code azure-spring-data-cosmos}</li>
 *   <li>DynamoDB    {@code spring-cloud-aws-dynamodb}</li>
 *   <li>Couchbase   {@code spring-boot-starter-data-couchbase}</li>
 *   <li>Firestore   {@code spring-cloud-gcp-data-firestore}</li>
 * </ul>
 *
 * Tables and schema migrations are explicitly out of scope here.
 */
public class Repositories {

    private static final Logger log = LoggerFactory.getLogger(Repositories.class);

    // ─── Device reports — immutable history ─────────────────────────────────
    @Component
    public static class DeviceReports {
        private final Map<String, Map<String, Object>> store = new ConcurrentHashMap<>();

        public void insert(Map<String, Object> doc) {
            if (doc == null) return;
            Object id = doc.get("_id");
            if (id == null) return;
            // In production: collection.insertOne(doc) — keyed by report id so
            // re-deliveries are idempotent.
            store.put(id.toString(), doc);
            log.info("insert device_report {} (role={})", id, doc.get("role"));
        }

        public Map<String, Object> get(String id) { return store.get(id); }
        public List<Map<String, Object>> all() { return new ArrayList<>(store.values()); }
        public void clear() { store.clear(); }
    }

    // ─── Devices — per-device latest state ──────────────────────────────────
    @Component
    public static class Devices {
        private final Map<String, Map<String, Object>> store = new ConcurrentHashMap<>();

        public boolean exists(String deviceId) {
            // Real implementations might keep a known-device cache, or do
            // collection.findOne({_id: deviceId}, projection: {_id: 1}). For
            // the stand-in we accept any non-empty id so the worker is exercised.
            return deviceId != null && !deviceId.isEmpty();
        }

        public void upsertLatest(String deviceId, Map<String, Object> patch) {
            if (deviceId == null || patch == null) return;
            store.compute(deviceId, (k, existing) -> {
                Map<String, Object> next = existing == null ? new LinkedHashMap<>() : new LinkedHashMap<>(existing);
                next.putAll(patch);
                next.put("_id", deviceId);
                return next;
            });
            // Production:
            //   mongo.upsert(
            //     Query.query(Criteria.where("_id").is(deviceId)),
            //     new Update().set(...patch...),
            //     "devices");
            log.info("upsert device {}", deviceId);
        }

        /**
         * Presence-only touch — a secondary's primary saw it during this
         * reporting window. The tag's AUTHORITATIVE sensor data lands in the
         * tag's own standalone report; do not overwrite sensor fields here.
         */
        public void touchObservation(String secondaryId,
                                     String primaryDeviceId,
                                     Integer rssi,
                                     String reportType,
                                     String time) {
            if (secondaryId == null) return;
            Map<String, Object> lastSeenBy = new LinkedHashMap<>();
            lastSeenBy.put("primaryDeviceId", primaryDeviceId);
            if (rssi != null) lastSeenBy.put("rssi", rssi);              // nullable
            if (reportType != null) lastSeenBy.put("reportType", reportType);
            if (time != null) lastSeenBy.put("time", time);

            store.compute(secondaryId, (k, existing) -> {
                Map<String, Object> next = existing == null ? new LinkedHashMap<>() : new LinkedHashMap<>(existing);
                next.put("_id", secondaryId);
                next.put("lastSeenBy", lastSeenBy);
                return next;
            });
            log.info("touch device {} seen by primary {}", secondaryId, primaryDeviceId);
        }

        /** Back-compat alias preserved for callers that hand in a JsonNode envelope. */
        public void updateLatest(String deviceId, JsonNode event) {
            Map<String, Object> patch = new LinkedHashMap<>();
            patch.put("lastEvent", event);
            upsertLatest(deviceId, patch);
        }

        public Map<String, Object> get(String id) { return store.get(id); }
        public List<Map<String, Object>> all() { return new ArrayList<>(store.values()); }
        public void clear() { store.clear(); }
    }

    // ─── Shipments — events appended; lookup by bound device/asset ──────────
    @Component
    public static class Shipments {
        private final Map<String, Map<String, Object>> store = new ConcurrentHashMap<>();

        @SuppressWarnings("unchecked")
        public void appendEvent(String shipmentId, Object event) {
            if (shipmentId == null) return;
            store.compute(shipmentId, (k, existing) -> {
                Map<String, Object> doc = existing == null
                        ? newShipmentDoc(shipmentId) : new LinkedHashMap<>(existing);
                ((List<Object>) doc.get("events")).add(event);
                return doc;
            });
            log.info("append event → shipment {}", shipmentId);
        }

        /**
         * Return the ID of the shipment associated with EITHER the device's
         * name (which may hold a shipment ID or name) OR the device's id
         * (if the device is assigned to an active shipment). The worker's
         * hottest read — cache aggressively in production.
         *
         * <p>Real implementation — a single indexed SQL statement:
         * <pre>
         *   SELECT id FROM shipments
         *   WHERE status IN ('ready', 'inProgress')
         *     AND (name = :name OR id = :name OR id IN (
         *       SELECT shipment_id FROM shipment_devices WHERE device_id = :id
         *     ))
         *   LIMIT 1;
         * </pre>
         */
        public String findIdByDeviceNameOrId(String deviceName, String deviceId) {
            log.info("lookup shipment by deviceName={} OR deviceId={}", deviceName, deviceId);
            for (Map.Entry<String, Map<String, Object>> entry : store.entrySet()) {
                if (deviceName != null && asList(entry.getValue().get("boundDeviceNames")).contains(deviceName)) return entry.getKey();
                if (deviceId != null && asList(entry.getValue().get("boundDeviceIds")).contains(deviceId)) return entry.getKey();
            }
            return null;
        }

        public String findIdByAssetNameOrId(String assetName, String assetId) {
            log.info("lookup shipment by assetName={} OR assetId={}", assetName, assetId);
            for (Map.Entry<String, Map<String, Object>> entry : store.entrySet()) {
                if (assetName != null && asList(entry.getValue().get("boundAssetNames")).contains(assetName)) return entry.getKey();
                if (assetId != null && asList(entry.getValue().get("boundAssetIds")).contains(assetId)) return entry.getKey();
            }
            return null;
        }

        public void seed(String shipmentId, Map<String, Object> overlay) {
            Map<String, Object> doc = newShipmentDoc(shipmentId);
            doc.putAll(overlay);
            store.put(shipmentId, doc);
        }

        public Map<String, Object> get(String id) { return store.get(id); }
        public void clear() { store.clear(); }

        private static Map<String, Object> newShipmentDoc(String id) {
            Map<String, Object> doc = new LinkedHashMap<>();
            doc.put("_id", id);
            doc.put("events", new ArrayList<>());
            return doc;
        }

        @SuppressWarnings("unchecked")
        private static List<Object> asList(Object value) {
            if (value instanceof List) return (List<Object>) value;
            return List.of();
        }
    }

    // ─── Assets — per-asset latest state ────────────────────────────────────
    @Component
    public static class Assets {
        private final Map<String, Map<String, Object>> store = new ConcurrentHashMap<>();

        public boolean exists(String assetId) {
            return assetId != null && !assetId.isEmpty();
        }

        public void updateLatest(String assetId, JsonNode event) {
            if (assetId == null) return;
            Map<String, Object> doc = new HashMap<>();
            doc.put("_id", assetId);
            doc.put("lastEvent", event);
            store.put(assetId, doc);
            log.info("upsert asset {}", assetId);
        }

        public Map<String, Object> get(String id) { return store.get(id); }
        public void clear() { store.clear(); }
    }
}

File: examples/csharp/Repositories.cs
using System.Collections.Concurrent;
using System.Text.Json;

namespace LocoAware.Webhook;

/// <summary>
/// In-memory stand-ins for a document store. The shape mirrors what you'd
/// push into MongoDB / DocumentDB / CosmosDB / Couchbase / Firestore. The
/// point of the examples is to show the SHAPE of the documents — the actual
/// driver is one swap away.
///
/// <para>Production replacements:
/// <list type="bullet">
///   <item><description>MongoDB:    <c>MongoDB.Driver</c> — <c>collection.UpdateOneAsync(filter, update, new UpdateOptions { IsUpsert = true })</c></description></item>
///   <item><description>CosmosDB:   <c>Microsoft.Azure.Cosmos</c> — <c>container.UpsertItemAsync(doc)</c></description></item>
///   <item><description>DynamoDB:   <c>AWSSDK.DynamoDBv2</c></description></item>
///   <item><description>Couchbase:  <c>Couchbase.Extensions.DependencyInjection</c></description></item>
///   <item><description>Firestore:  <c>Google.Cloud.Firestore</c></description></item>
/// </list>
/// Tables and schema migrations are explicitly out of scope here.</para>
/// </summary>
public class DeviceReportsRepository
{
    private readonly ILogger<DeviceReportsRepository> _log;
    private readonly ConcurrentDictionary<string, IDictionary<string, object?>> _store = new();

    public DeviceReportsRepository(ILogger<DeviceReportsRepository> log) => _log = log;

    public virtual Task InsertAsync(IDictionary<string, object?> doc)
    {
        if (doc.TryGetValue("_id", out var idObj) && idObj is string id && !string.IsNullOrEmpty(id))
        {
            // In production: await collection.InsertOneAsync(doc) — keyed by
            // report id so re-deliveries are idempotent.
            _store[id] = doc;
            _log.LogInformation("insert device_report {Id} (role={Role})", id, doc.TryGetValue("role", out var r) ? r : "unknown");
        }
        return Task.CompletedTask;
    }

    public virtual IDictionary<string, object?>? Get(string id) => _store.TryGetValue(id, out var d) ? d : null;
    public virtual IReadOnlyCollection<IDictionary<string, object?>> All() => _store.Values.ToList();
    public virtual void Clear() => _store.Clear();
}

public class ShipmentsRepository
{
    private readonly ILogger<ShipmentsRepository> _log;
    private readonly ConcurrentDictionary<string, IDictionary<string, object?>> _store = new();

    public ShipmentsRepository(ILogger<ShipmentsRepository> log) => _log = log;

    public virtual Task AppendEventAsync(string shipmentId, JsonElement @event)
        => AppendEventAsync(shipmentId, (object)@event);

    public virtual Task AppendEventAsync(string shipmentId, object @event)
    {
        if (string.IsNullOrEmpty(shipmentId)) return Task.CompletedTask;
        _store.AddOrUpdate(
            shipmentId,
            _ =>
            {
                var doc = NewDoc(shipmentId);
                ((List<object?>)doc["events"]!).Add(@event);
                return doc;
            },
            (_, existing) =>
            {
                ((List<object?>)existing["events"]!).Add(@event);
                return existing;
            });
        _log.LogInformation("append event → shipment {ShipmentId}", shipmentId);
        return Task.CompletedTask;
    }

    /// <summary>
    /// Return the shipment ID associated with EITHER the device's name (which
    /// may hold a shipment ID or name) OR the device's id (if the device is
    /// currently assigned to an active shipment). The worker's hottest read
    /// — cache aggressively in production.
    ///
    /// Real implementation — one indexed SQL statement:
    /// <code>
    ///   SELECT id FROM shipments
    ///   WHERE status IN ('ready', 'inProgress')
    ///     AND (name = @name OR id = @name OR id IN (
    ///       SELECT shipment_id FROM shipment_devices WHERE device_id = @id
    ///     ))
    ///   LIMIT 1;
    /// </code>
    /// </summary>
    public virtual Task<string?> FindIdByDeviceNameOrIdAsync(string? deviceName, string? deviceId)
    {
        _log.LogInformation("lookup shipment by deviceName={DeviceName} OR deviceId={DeviceId}", deviceName, deviceId);
        foreach (var kv in _store)
        {
            if (deviceName is not null && ListContains(kv.Value, "boundDeviceNames", deviceName)) return Task.FromResult<string?>(kv.Key);
            if (deviceId is not null && ListContains(kv.Value, "boundDeviceIds", deviceId)) return Task.FromResult<string?>(kv.Key);
        }
        return Task.FromResult<string?>(null);
    }

    public virtual Task<string?> FindIdByAssetNameOrIdAsync(string? assetName, string? assetId)
    {
        _log.LogInformation("lookup shipment by assetName={AssetName} OR assetId={AssetId}", assetName, assetId);
        foreach (var kv in _store)
        {
            if (assetName is not null && ListContains(kv.Value, "boundAssetNames", assetName)) return Task.FromResult<string?>(kv.Key);
            if (assetId is not null && ListContains(kv.Value, "boundAssetIds", assetId)) return Task.FromResult<string?>(kv.Key);
        }
        return Task.FromResult<string?>(null);
    }

    public virtual void Seed(string id, IDictionary<string, object?> overlay)
    {
        var doc = NewDoc(id);
        foreach (var kv in overlay) doc[kv.Key] = kv.Value;
        _store[id] = doc;
    }

    public virtual IDictionary<string, object?>? Get(string id) => _store.TryGetValue(id, out var d) ? d : null;
    public virtual void Clear() => _store.Clear();

    private static IDictionary<string, object?> NewDoc(string id) => new Dictionary<string, object?>(StringComparer.Ordinal)
    {
        ["_id"] = id,
        ["events"] = new List<object?>(),
    };

    private static bool ListContains(IDictionary<string, object?> doc, string key, string value)
        => doc.TryGetValue(key, out var v)
           && v is System.Collections.IEnumerable list
           && list.Cast<object?>().Any(x => x is string s && s == value);
}

public class DevicesRepository
{
    private readonly ILogger<DevicesRepository> _log;
    private readonly ConcurrentDictionary<string, IDictionary<string, object?>> _store = new();

    public DevicesRepository(ILogger<DevicesRepository> log) => _log = log;

    public virtual Task<bool> ExistsAsync(string deviceId)
    {
        // Real implementations might keep a known-device cache, or do
        // collection.FindAsync(filter). For the stand-in we accept any
        // non-empty id so the worker is exercised.
        return Task.FromResult(!string.IsNullOrEmpty(deviceId));
    }

    public virtual Task UpsertLatestAsync(string deviceId, IDictionary<string, object?> patch)
    {
        if (string.IsNullOrEmpty(deviceId)) return Task.CompletedTask;
        _store.AddOrUpdate(
            deviceId,
            _ =>
            {
                var doc = new Dictionary<string, object?>(patch, StringComparer.Ordinal) { ["_id"] = deviceId };
                return doc;
            },
            (_, existing) =>
            {
                foreach (var kv in patch) existing[kv.Key] = kv.Value;
                existing["_id"] = deviceId;
                return existing;
            });
        // Production:
        //   await collection.UpdateOneAsync(
        //     Builders<BsonDocument>.Filter.Eq("_id", deviceId),
        //     Builders<BsonDocument>.Update.Set(...),
        //     new UpdateOptions { IsUpsert = true });
        _log.LogInformation("upsert device {DeviceId}", deviceId);
        return Task.CompletedTask;
    }

    /// <summary>
    /// Presence-only touch — a secondary's primary saw it during this
    /// reporting window. The tag's AUTHORITATIVE sensor data lands in the
    /// tag's own standalone report; do not overwrite sensor fields here.
    /// </summary>
    public virtual Task TouchObservationAsync(string secondaryId, string? primaryDeviceId, int? rssi, string? reportType, string? time)
    {
        if (string.IsNullOrEmpty(secondaryId)) return Task.CompletedTask;
        var lastSeenBy = new Dictionary<string, object?>(StringComparer.Ordinal)
        {
            ["primaryDeviceId"] = primaryDeviceId,
        };
        if (rssi.HasValue) lastSeenBy["rssi"] = rssi.Value;          // nullable
        if (reportType is not null) lastSeenBy["reportType"] = reportType;
        if (time is not null) lastSeenBy["time"] = time;

        _store.AddOrUpdate(
            secondaryId,
            _ => new Dictionary<string, object?>(StringComparer.Ordinal)
            {
                ["_id"] = secondaryId,
                ["lastSeenBy"] = lastSeenBy,
            },
            (_, existing) =>
            {
                existing["lastSeenBy"] = lastSeenBy;
                existing["_id"] = secondaryId;
                return existing;
            });
        _log.LogInformation("touch device {SecondaryId} seen by primary {PrimaryId}", secondaryId, primaryDeviceId);
        return Task.CompletedTask;
    }

    /// <summary>Back-compat alias for callers handing in a raw envelope.</summary>
    public virtual Task UpdateLatestAsync(string deviceId, JsonElement @event)
    {
        var patch = new Dictionary<string, object?>(StringComparer.Ordinal) { ["lastEvent"] = @event };
        return UpsertLatestAsync(deviceId, patch);
    }

    public virtual IDictionary<string, object?>? Get(string id) => _store.TryGetValue(id, out var d) ? d : null;
    public virtual void Clear() => _store.Clear();
}

public class AssetsRepository
{
    private readonly ILogger<AssetsRepository> _log;
    private readonly ConcurrentDictionary<string, IDictionary<string, object?>> _store = new();

    public AssetsRepository(ILogger<AssetsRepository> log) => _log = log;

    public virtual Task<bool> ExistsAsync(string assetId) => Task.FromResult(!string.IsNullOrEmpty(assetId));

    public virtual Task UpdateLatestAsync(string assetId, JsonElement @event)
    {
        if (string.IsNullOrEmpty(assetId)) return Task.CompletedTask;
        _store[assetId] = new Dictionary<string, object?>(StringComparer.Ordinal)
        {
            ["_id"] = assetId,
            ["lastEvent"] = @event,
        };
        _log.LogInformation("upsert asset {AssetId}", assetId);
        return Task.CompletedTask;
    }

    public virtual IDictionary<string, object?>? Get(string id) => _store.TryGetValue(id, out var d) ? d : null;
    public virtual void Clear() => _store.Clear();
}
